framework/stats/stats.cpp

```diff
@@ -0,0 +1,273 @@
+	providers.emplace_back(std::make_unique<VulkanStatsProvider>(device, stats, sampling_config, num_framebuffers));
+
+	// In continuous sampling mode we still need to update the frame times as if we are polling
+	// Store the frame time provider here so we can easily access it later.
+	frame_time_provider = providers[0].get();
+
+	for (const auto &stat : requested_stats)
+	{
+		counters[stat] = std::vector<float>(buffer_size, 0);
+	}
+
+	if (sampling_config.mode == CounterSamplingMode::Continuous)
+	{
+		// Start a thread for continuous sample capture
+		worker_thread = std::thread([this] {
+			continuous_sampling_worker(stop_worker->get_future());
+		});
+
+		// Reduce smoothing for continuous sampling
+		alpha_smoothing = 0.6f;
+	}
+}
+
+Stats::~Stats()
+{
+	if (stop_worker)
+	{
+		stop_worker->set_value();
+	}
+
+	if (worker_thread.joinable())
+	{
+		worker_thread.join();
+	}
+}
+
+void Stats::resize(const size_t width)
+{
+	// The circular buffer size will be 1/16th of the width of the screen
+	// which means every sixteen pixels represent one graph value
+	size_t buffers_size = width >> 4;
+
+	for (auto &counter : counters)
+	{
+		counter.second.resize(buffers_size);
+		counter.second.shrink_to_fit();
+	}
+}
+
+bool Stats::is_available(const StatIndex index) const
+{
+	for (const auto &p : providers)
+		if (p->is_available(index))
+			return true;
+	return false;
+}
+
+static void add_smoothed_value(std::vector<float> &values, float value, float alpha)
+{
+	assert(values.size() >= 2 && "Buffers size should be greater than 2");
+
+	if (values.size() == values.capacity())
+	{
+		// Shift values to the left to make space at the end and update counters
+		std::rotate(values.begin(), values.begin() + 1, values.end());
+	}
+
+	// Use an exponential moving average to smooth values
+	values.back() = value * alpha + *(values.end() - 2) * (1.0f - alpha);
+}
+
+void Stats::update(float delta_time, uint32_t active_frame_idx)
+{
+	switch (sampling_config.mode)
+	{
+		case CounterSamplingMode::Polling:
+		{
+			StatsProvider::Sample sample;
+
+			for (auto &p : providers)
+			{
+				auto s = p->sample(delta_time, active_frame_idx);
+				sample.insert(s.begin(), s.end());
+			}
+			push_sample(sample);
+			break;
+		}
+		case CounterSamplingMode::Continuous:
+		{
+			// Check that we have no pending samples to be shown
+			if (pending_samples.size() == 0)
+			{
+				std::unique_lock<std::mutex> lock(continuous_sampling_mutex);
+				if (!should_add_to_continuous_samples)
+				{
+					// If we have no pending samples, we let the worker thread
+					// capture samples for the next frame
+					should_add_to_continuous_samples = true;
+				}
+				else
+				{
+					// The worker thread has captured a frame, so we stop it
+					// and read the samples
+					should_add_to_continuous_samples = false;
+					pending_samples                  = continuous_samples;
+					continuous_samples.clear();
+				}
+			}
+
+			if (pending_samples.size() == 0)
```

Comment:
Throughout the project we use curly braces in our conditionals - even if they are one liners. I understand that by the guidelines we use that this is legal, however it's making this code seem inconsistent with the rest of the project. Seeing as the original code for these files had the braces could we keep it that way?

Reply:
Nitpick

Reply:
I can go through and change all these if that's really necessary.
As you say, the guidelines don't specify this; I checked before I did it.
This is also done in several other places in the samples.

If you really want me to change all these, can I first ask that this is added to the coding style rules and the other places get fixed up too.